src/jit/codegencommon.cpp

```diff
@@ -4888,6 +4888,82 @@ void            CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg,
 #pragma warning(pop)
 #endif
 
+
+/*****************************************************************************
+* If any Vector3 args are on stack and they are not pass-by-ref, the upper 32bits
+* must be cleared to zeroes. The native compiler doesn't clear the upper bits
+* and there is no way to know if the caller is native or not. So, the upper
+* 32 bits of Vector argument on stack are always cleared to zero.
+*/
+#ifdef FEATURE_SIMD
+void    CodeGen::genClearStackVec3ArgUpperBits()
+{
+#ifdef DEBUG
+    if (verbose)
+        printf("*************** In genClearStackVec3ArgUpperBits()\n");
+#endif
+
+#if !defined(_TARGET_XARCH_)
+    return;
+#endif // !_TARGET_XARCH_
```

Comment:
Yes, `genFnPrologCalleeRegArgs` is horribly complicated, and one of the structs-related work items is to refactor this. In looking at it once again, it seems that it's not really inefficient to do them separately.
However, I think that's simply due to the inherent limitations of the SSE/AVX instruction set.
Here are the two cases of interest:
Case 1: The arg is passed on the stack, and then "homed" into a register. In this case, you will clear the upper bits on the stack in `genClearStackVec3ArgUpperBits()`, and then load it into a register in `genEnregisterIncomingStackArgs()`. That may be the most efficient (the alternative would be to do two loads and a shuffle, versus a store and a load).
Case 2: The arg is passed in a register and homed onto the stack. In this case you will issue two shuffles in `genClearStackVec3ArgUpperBits()` and then a 128-bit store in `genEnregisterIncomingStackArgs()`. This is probably more efficient than storing the 8-byte and 4-byte pieces separately, though I'm not certain about that.
The issue, to me, is that this adds yet another place where the args are messed with. I think that is more confusing than splitting the logic for the reg vs. stack case.
In any event, the comment starting with `// Assumption 1:` in `genFnPrologCalleeRegArgs()` needs to be updated.